File: ios/Sources/VaultSync/Domain/UseCases/DeleteDocumentUseCase.swift
import Foundation

public final class DeleteDocumentUseCase: Sendable {
    private let documentRepository: any DocumentRepository
    private let syncRepository: any SyncRepository
    private let fileStorage: any FileStorage

    public init(
        documentRepository: any DocumentRepository,
        syncRepository: any SyncRepository,
        fileStorage: any FileStorage
    ) {
        self.documentRepository = documentRepository
        self.syncRepository = syncRepository
        self.fileStorage = fileStorage
    }

    public func execute(documentId: String) async throws {
        guard let document = try await documentRepository.getDocument(id: documentId) else {
            throw AppError.notFoundError("Document \(documentId) not found.")
        }

        if await fileStorage.fileExists(at: document.localPath) {
            _ = try await fileStorage.deleteFile(at: document.localPath)
        }

        try await documentRepository.deleteDocument(id: documentId)

        let syncOp = SyncOperation(
            id: UUID().uuidString,
            documentId: documentId,
            operationType: .delete,
            status: .pending,
            startedAt: Date()
        )
        try await syncRepository.recordOperation(syncOp)
    }
}

File: android/app/src/main/java/com/vaultsync/domain/usecase/DeleteDocumentUseCase.kt
package com.vaultsync.domain.usecase

import com.vaultsync.domain.model.AppError
import com.vaultsync.domain.model.OperationType
import com.vaultsync.domain.model.SyncOperation
import com.vaultsync.domain.repository.DocumentRepository
import com.vaultsync.domain.repository.FileStorage
import com.vaultsync.domain.repository.SyncRepository
import java.util.UUID

class DeleteDocumentUseCase(
    private val documentRepository: DocumentRepository,
    private val syncRepository: SyncRepository,
    private val fileStorage: FileStorage
) {
    suspend operator fun invoke(documentId: String) {
        val document = documentRepository.getDocument(documentId)
            ?: throw AppError.NotFoundError("Document with ID $documentId not found.")

        if (fileStorage.fileExists(document.localPath)) {
            fileStorage.deleteFile(document.localPath)
        }

        documentRepository.deleteDocument(documentId)

        val syncOp = SyncOperation(
            id = UUID.randomUUID().toString(),
            documentId = documentId,
            operationType = OperationType.DELETE,
            status = com.vaultsync.domain.model.OperationStatus.PENDING,
            startedAt = System.currentTimeMillis()
        )
        syncRepository.recordOperation(syncOp)
    }
}
